soccer/src/soccer/planning/planner_node.cpp

```diff
@@ -27,6 +44,76 @@ PlannerNode::PlannerNode()
     }
 }
 
+rclcpp_action::GoalResponse PlannerNode::handle_goal(const rclcpp_action::GoalUUID& uuid,
+                                                     std::shared_ptr<const RobotMove::Goal> goal) {
+    (void)uuid;
+
+    // TODO(p-nayak): REJECT duplicate goal requests so we aren't constantly replanning them
+
+    // planning::MotionCommand motion_command_ = goal->robot_intent.motion_command;
+
+    SPDLOG_ERROR("robot id {} sent goal", goal->robot_intent.robot_id);
+    return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
+}
+
+rclcpp_action::CancelResponse PlannerNode::handle_cancel(
+    const std::shared_ptr<GoalHandleRobotMove> goal_handle) {
+    (void)goal_handle;
+    std::shared_ptr<const RobotMove::Goal> goal = goal_handle->get_goal();
+    SPDLOG_ERROR("action server canceled goal from robot: {}", goal->robot_intent.robot_id);
+    return rclcpp_action::CancelResponse::ACCEPT;
+}
+
+void PlannerNode::handle_accepted(const std::shared_ptr<GoalHandleRobotMove> goal_handle) {
+    // this needs to return quickly to avoid blocking the executor, so spin up a new thread
+    // execute() will block (loop) until completion (either success or canceled by client)
+    using namespace std::placeholders;
+    std::thread{std::bind(&PlannerNode::execute, this, _1), goal_handle}.detach();
+}
+
+void PlannerNode::execute(const std::shared_ptr<GoalHandleRobotMove> goal_handle) {
+    // rate-limit loop to 1ms per iteration
+    rclcpp::Rate loop_rate(1);
+
+    // create ptrs to Goal, Result objects per ActionServer API
+    std::shared_ptr<const RobotMove::Goal> goal = goal_handle->get_goal();
+    std::shared_ptr<RobotMove::Result> result = std::make_shared<RobotMove::Result>();
+
+    // get correct PlannerForRobot object for this robot_id
+    int robot_id = goal->robot_intent.robot_id;
+    std::shared_ptr<PlannerForRobot> my_robot_planner = robots_planners_[robot_id];
+
+    // loop until goal is done (SUCCEEDED or CANCELED)
+    bool not_done_yet = true;  // yes, this is necessary (compiler optimizes out while(true))
+    while (not_done_yet) {
```

Comment:
```suggestion
    for (;;) {
```
It worked for me like this. I think it is more clear this way that the loop will only end on external factors.

Reply:
oh sweet, good to know